decision_service/rlclientlib/factory_resolver.cc

```diff
@@ -0,0 +1,70 @@
+#include "factory_resolver.h"
+#include <type_traits>
+#include "constants.h"
+#include "restapi_data_transport.h"
+#include "vw_model.h"
+#include "err_constants.h"
+
+namespace reinforcement_learning { 
+  namespace m = model_management;
+  namespace u = utility;
+  // For proper static intialization 
+  // Check https://en.wikibooks.org/wiki/More_C++_Idioms/Nifty_Counter for explanation
+  static int init_guard;  // guranteed to be zero when loaded
+
+  // properly aligned memory for the factory object 
+  template <typename T>
+  using natural_align = std::aligned_storage<sizeof(T), alignof ( T )>;
```

Comment:
why is this needed?

Reply:
It looks like a ref-count for the purposes of initializing/destructing the singleton. My question is - why not use a shared pointer or similar for this?

Reply:
normally new allocates properly aligned memory and invokes the constructor.  This template is used to make sure that statically memory allocated for factory objects have the right alignment for that object.

---
In reply to: [193486212](https://github.com/JohnLangford/vowpal_wabbit/pull/1495#discussion_r193486212) [](ancestors = 193486212)

Reply:
This is a technique used to initialize static objects because the initialization order is not guaranteed across translation units.  This an issue for example if the users of this library decide to use this object inside their own global object.

For example std::cout uses this method for initializing.  

more details here: [https://en.wikibooks.org/wiki/More_C++_Idioms/Nifty_Counter](https://en.wikibooks.org/wiki/More_C++_Idioms/Nifty_Counter)


---
In reply to: [193488308](https://github.com/JohnLangford/vowpal_wabbit/pull/1495#discussion_r193488308) [](ancestors = 193488308)